go/config/config_test.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2019 The Grafeas Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func defaultServerConfig() *ServerConfig {
+	return &ServerConfig{
+		Address:            "0.0.0.0:8080",
+		CertFile:           "",
+		KeyFile:            "",
+		CAFile:             "",
+		CORSAllowedOrigins: nil,
+	}
+}
+
+func userServerConfig() *ServerConfig {
+	return &ServerConfig{
+		Address:  "0.0.0.0:8081",
+		CertFile: "abc",
+		KeyFile:  "def",
+		CAFile:   "ghi",
+		CORSAllowedOrigins: []string{
+			"http://example.com",
+			"https://somewhere.else.com",
+		},
+	}
+}
+
+var userConfig_memstore_yaml = []byte(`
+grafeas:
+  api:
+    address: "0.0.0.0:8081"
+    certfile: abc
+    keyfile: def
+    cafile:  ghi
+    cors_allowed_origins:
+      - "http://example.com"
+      - "https://somewhere.else.com"
+  storage_type: "memstore"
+`)
+
+func userEmbeddedConfig() *EmbeddedStoreConfig {
+	return &EmbeddedStoreConfig{
+		Path: "/some/path",
+	}
+}
+
+var userConfig_embedded_yaml = []byte(`
+grafeas:
+  api:
+    address: "0.0.0.0:8081"
+    certfile: abc
+    keyfile: def
+    cafile:  ghi
+    cors_allowed_origins:
+      - "http://example.com"
+      - "https://somewhere.else.com"
+  storage_type: "embedded"
+  embedded: 
+    path: /some/path
+`)
+
+func TestLoadConfig_ReturnsDefaultConfig_NoInput(t *testing.T) {
```

Comment:
Ping on this. Table tests are the canonical and idiomatic way to parameterize tests in Go and makes it cleaner and easier to read. It looks like this can be turned into a table test.

Reply:
Actually, turning this into a table test doesn't really work.  There are a couple of issues that I see:

1. Configuration for different storage types are defined as completely different structs, e.g. `PgSQLConfig`, `EmbeddedStoreConfig`, with nothing in common between them.  A table-driven test needs to allow the user to specify what the input config file contents look like and also what the actual resulting struct should be, e.g.:

```
type LoadConfigTestConfiguration struct {
	// name of the test
	description string
	// function that will return the desired input configuration, nil for no input config
	inputConfigFileContentsProvider func(*testing.T) []byte
	// function returning expected ServerConfig
	expectedServerConfig func(*testing.T) *ServerConfig
	// expected StorageType name
	expectedStorageTypeName string
	// function returning expected StorageConfig, nil to assert that you wish to check for no config returned
	expectedStorageConfig func(*testing.T) StorageConfiguration
}

var configTests = []LoadConfigTestConfiguration{
	{ "returns default config when no input", nil,  defaultServerConfig, "memstore", nil},
	{ "returns correct user-supplied memstore config", userConfig_memstore_yaml,  userServerConfig, "memstore", nil},
	{ "returns correct user-supplied embedded config", userConfig_embedded_yaml,  userServerConfig, "embedded", userEmbeddedConfig},
}

// expected configuration object that should be created for embedded storage
func userEmbeddedConfig(t *testing.T) StorageConfiguration {
	t.Helper()
	return &EmbeddedStoreConfig {
		Path: "/some/path",
	}
}


// within a test runner we need to compare the config that was created from the YAML with the expected StorageConfiguration

                        // ...
			cfg, err := LoadConfig(fileName) // load config from a temp file, returning (*GrafeasConfig, err)
                        // ...

                        // test is an instance of LoadConfigTestConfiguration
			if test.expectedStorageConfig == nil {
                               // this branch tests for memstore, where no configuration should be created
				if cfg.StorageConfig != nil {
					t.Errorf("storage configuration is not nil")
				}
			} else {
                                // there should be some form of config created, so test it is what was expected
				if !cmp.Equal(*cfg.StorageConfig, test.expectedStorageConfig(t)) {
					t.Errorf("Values in storage configuration are not correct\n%s", cmp.Diff(*cfg.StorageConfig, test.expectedStorageConfig(t)))
				}
			}
```

We get the following error when using `cmp.Diff()`
```
 Values in storage configuration are not correct
              interface{}(
            -   &map[string]interface{}{"path": string("/some/path")},
            +   &config.EmbeddedStoreConfig{Path: "/some/path"},
              )
```
So, the issue boils down to the fact that we are comparing an `interface{}` with a specific type, even though that specific type is cast to an `interface{}`.  Note that currently each existing test actually instantiates an instance of the expected type from the created `interface{}` and compares that instance with the expected specific type.  This is because each test is aware of what it is trying to test for, rather than being a one-size-fits-all generic test function.  I'm not sure how to work around this - we actually want to compare between `EmbeddedStoreConfig` and `EmbeddedStoreConfig` to make the test pass.  If we turn the `expectedStorageConfig` into a raw `interface{}` via marshalling to JSON and back again, we get a similar problem on comparison, where the fields are not capitalized:

```
Values in storage configuration are not correct
              map[string]interface{}{
            +   "Path": string("/some/path"),
            -   "path": string("/some/path"),
              }
```

2.  The test runner now starts to have logic in it to cater for different values supplied within the test table (e.g. , which in my opinion is a serious code smell.

If you've got ideas as to how you'd like this done then let me know and I'll try to find some time to implement it, but I see you approved the code review so I'll mark this conversation as resolved for now so that GitHub doesn't flag this point as outstanding.